csrc/python_frontend/fusion_cache.cpp

```diff
@@ -30,15 +31,21 @@ Fusion* FusionSchedules::preschedFusion() {
   return fusion;
 }
 
-TrieNode::TrieNode(RecordFunctor* rec, size_t _fusion_id)
-    : record(rec), children(), fusion_id(_fusion_id), visits(0) {}
+TrieNode::TrieNode(RecordFunctor* rec, TrieNode* _parent, size_t _fusion_id)
+    : record(rec),
+      children(),
+      fusion_id(_fusion_id),
+      visits(0),
+      parent(_parent),
+      trie_node_lock() {}
 
 bool TrieNode::isTerminal() const {
   return (record.get()->recordType() == RecordType::End);
 }
 
 FusionCache* FusionCache::get(size_t max_fusions) {
-  std::lock_guard<std::mutex> guard(fusion_cache_lock);
+  FUSER_PERF_SCOPE("FusionCache::get");
+  std::lock_guard<std::mutex> guard(singleton_lock_);
```

Comment:
the lock is released after this function. But we do return non-cont pointer to the singleton after releasing the lock. This doesn't feel right.

Reply:
I can't make the `FusionCache` cache object `const` at this time given that the `FusionDefinition` holds a pointer and can ask to create nodes from the pointer.  Leaving this issue for now.